### DISCUSSION TITLE
VL53L3CX Tof sensor

### DISCUSSION THREAD

**Discussion by jet689** [micropython/micropython#14433] [Hardware & Peripherals]
I have been aching my brain for the last couple of days. I can't figure out how to use the VL53L3CX ToF sensor. The actual datasheet doesn't include much for how to use its i2c. There is a stm32 cube API for it but I really haven't a clue about that. That is about as much as I have figured out. 

Ps: I am using a pi pico.

**Comment by ricksorensen** [comment]
A quick search gives many possibilities for starting with the chip (which I do not have.)  Here is one from [awesome micropython](https://awesome-micropython.com/)

[https://github.com/mp-extras/vl53l5cx](https://github.com/mp-extras/vl53l5cx)

**Reply by jet689** [reply]
thanks for the quick reply. the one you sent me is for the 5cx version do you have something similar for the 3cx which is the chip I have?

**Reply by ricksorensen** [reply]
I don't have any particular solutions for you (not having used either chip)- maybe expand your internet searches to include CircuitPy and Arduino.  It looks like the actual messaging specification documents are not readily available, so you may need to look at source code.

Good luck

**Reply by FrgyCZ** [reply]
Hi, I don't know if it's still relevant one year later or if it works on rpi pico but you may give it a shot. I have tested it on rpi zero 2 with great success. I open-sourced it on github, its a python library that calls a c code that directly interacts with ST's VL53L3CX API.
https://github.com/FrgyCZ/VL53L3CX-python

**Comment by aliarifat** [comment]
The VL53L3CX ToF sensor can be a bit tricky to use because ST does not expose its full I²C register map in the datasheet. Instead, they provide a precompiled driver (Ultra Lite Driver) that must be used to initialize the sensor and retrieve distance measurements. This is why many beginners feel stuck when trying to directly communicate with the sensor using only raw I²C transactions on platforms like the Raspberry Pi Pico.

One of the easiest ways to get started is to use the Arduino core on the Pico along with a wrapper library such as Pololu’s VL53L3CX library. This avoids having to manually port ST’s driver and gives you quick access to ranging results with just a few lines of code. Alternatively, for those comfortable with C, you can port the official driver by replacing the low-level I²C read/write functions with Pico SDK equivalents. This offers more control but is a bit more involved.

For some additional inspiration, you can look at this [ToF Sensor PCB for obstacle avoidance](https://www.pcbway.com/project/shareproject/CrazyFlie_Obstacle_Avoidance_Deck.html)
 project, which shows how ToF sensors like the VL53 series are integrated into practical applications. It’s a great example of how these sensors are deployed in real-world robotics, and the same concepts can be applied to your Pico project.
